pool.go

```diff
@@ -0,0 +1,273 @@
+package connection
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type ConnectionFactoryFunc func(addr string) (*Connection, error)
+
+type Pool struct {
+	Factory ConnectionFactoryFunc
+	Addrs   []string
+	Opts    PoolOptions
+
+	mu          sync.Mutex
+	connections []*Connection
+	connIndex   uint32
+
+	done chan struct{}
+
+	// WaitGroup to wait for all reconnect goroutines return
+	wg sync.WaitGroup
+
+	isClosed bool
+}
+
+// Pool - provides connections to the clients. It removes the connection from
+// the pool if it was closed and in the background tries to create and etablish
+// new connection so the pool will be full.
+func NewPool(factory ConnectionFactoryFunc, addrs []string, options ...PoolOption) (*Pool, error) {
+	opts := GetDefaultPoolOptions()
+	for _, opt := range options {
+		if err := opt(&opts); err != nil {
+			return nil, fmt.Errorf("setting pool option: %v %w", opt, err)
+		}
+	}
+
+	return &Pool{
+		Factory:     factory,
+		Opts:        opts,
+		Addrs:       addrs,
+		done:        make(chan struct{}),
+		connections: make([]*Connection, 0),
+	}, nil
+}
+
+func (p *Pool) handleError(err error) {
+	if p.Opts.ErrorHandler == nil {
+		return
+	}
+
+	go p.Opts.ErrorHandler(err)
+}
+
+// Connect creates poll of connections by calling Factory method and connect them all
+func (p *Pool) Connect() error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if p.isClosed {
+		return errors.New("pool is closed")
+	}
+
+	// build connections
+	for _, addr := range p.Addrs {
+		conn, err := p.Factory(addr)
+		if err != nil {
+			return fmt.Errorf("creating connection for %s: %w", addr, err)
+		}
+
+		// set own handler when connection is closed
+		conn.SetOptions(ConnectionClosedHandler(p.handleClosedConnection))
+
+		err = conn.Connect()
+		if err != nil {
+			p.handleError(fmt.Errorf("connecting to %s: %w", conn.addr, err))
+			p.wg.Add(1)
+			go p.recreateConnection(conn)
+			continue
+		}
+
+		p.connections = append(p.connections, conn)
+	}
+
+	if len(p.connections) < p.Opts.MinConnections {
+		return fmt.Errorf("minimum %d connections is required, established: %d", p.Opts.MinConnections, len(p.connections))
+	}
+
+	return nil
+}
+
+// Connections returns copy of all connections from the pool
+func (p *Pool) Connections() []*Connection {
```

Comment:
This method seems odd to me, was it done for testing? Seems like this should remain encapsulated below the public API

Reply:
yeah. I plan to remove it soon. I don't think we will need it in the future, but we need it for now.